File: Phase-2/backend/app/models/task.py
"""
Task model and Pydantic schemas.

Defines SQLModel entity and request/response schemas for tasks.

@spec: specs/002-todo-backend-api/spec.md (FR-002, FR-003, FR-004, FR-005)
"""
from datetime import datetime, timezone
from typing import Optional
from uuid import UUID, uuid4

from sqlmodel import Field as SQLField, SQLModel
from pydantic import BaseModel, Field, field_validator, field_serializer


class Task(SQLModel, table=True):
    """
    Task database model.

    All queries MUST filter by user_id for multi-tenancy.

    @spec: specs/002-todo-backend-api/spec.md (FR-002, FR-010)
    """
    __tablename__ = "tasks"

    id: UUID = SQLField(default_factory=uuid4, primary_key=True, nullable=False)
    user_id: str = SQLField(nullable=False, index=True, max_length=255)
    title: str = SQLField(nullable=False, max_length=500)
    description: Optional[str] = SQLField(default=None, nullable=True)
    status: str = SQLField(default="pending", nullable=False, max_length=50, index=True)
    priority: Optional[str] = SQLField(default="medium", nullable=True, max_length=50)
    created_at: datetime = SQLField(default_factory=datetime.utcnow, nullable=False, index=True)
    updated_at: datetime = SQLField(default_factory=datetime.utcnow, nullable=False)


# ── Request / Response Pydantic Schemas ─────────────────────────────────────

class TaskCreateRequest(BaseModel):
    """Schema for creating a new task."""
    title: str = Field(..., min_length=1, max_length=500, description="Task title")
    description: Optional[str] = Field(None, max_length=10000, description="Task description")
    priority: Optional[str] = Field("medium", description="Task priority: low | medium | high")

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: str) -> str:
        v = v.strip()
        if not v:
            raise ValueError("Title cannot be empty or whitespace")
        return v

    @field_validator("description")
    @classmethod
    def validate_description(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            return v if v else None
        return None

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ("low", "medium", "high"):
            raise ValueError("priority must be one of: low, medium, high")
        return v


class TaskUpdateRequest(BaseModel):
    """Schema for updating an existing task."""
    title: Optional[str] = Field(None, min_length=1, max_length=500)
    description: Optional[str] = Field(None, max_length=10000)
    status: Optional[str] = Field(None, description="pending | in_progress | completed")
    priority: Optional[str] = Field(None, description="low | medium | high")

    @field_validator("title")
    @classmethod
    def validate_title(cls, v: Optional[str]) -> Optional[str]:
        if v is not None:
            v = v.strip()
            if not v:
                raise ValueError("Title cannot be empty or whitespace")
            return v
        return None

    @field_validator("status")
    @classmethod
    def validate_status(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ("pending", "in-progress", "completed"):
            raise ValueError("status must be one of: pending, in-progress, completed")
        return v

    @field_validator("priority")
    @classmethod
    def validate_priority(cls, v: Optional[str]) -> Optional[str]:
        if v is not None and v not in ("low", "medium", "high"):
            raise ValueError("priority must be one of: low, medium, high")
        return v


class TaskResponse(BaseModel):
    """Schema for task response."""
    id: UUID
    user_id: str
    title: str
    description: Optional[str]
    status: str
    priority: Optional[str]
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

    @field_serializer("created_at", "updated_at")
    def serialize_datetime(self, dt: datetime) -> str:
        """Always serialize datetimes as UTC ISO strings with Z suffix.

        Pydantic v2 serializes naive datetimes without timezone info,
        which Zod's z.string().datetime() rejects. We store naive UTC
        datetimes in the DB, so we just append 'Z' to signal UTC.
        """
        if dt.tzinfo is None:
            return dt.isoformat() + "Z"
        return dt.astimezone(timezone.utc).isoformat().replace("+00:00", "Z")


class TaskListResponse(BaseModel):
    """Schema for task list response."""
    status: str = "success"
    data: list[TaskResponse]
    meta: dict = Field(default_factory=lambda: {"total": 0})


class TaskSingleResponse(BaseModel):
    """Schema for single task response."""
    status: str = "success"
    data: TaskResponse
    message: Optional[str] = None


class ErrorResponse(BaseModel):
    """Schema for error responses."""
    status: str = "error"
    code: str
    message: str
    details: Optional[dict] = None

File: Phase-2/frontend/src/lib/schemas/task.ts
/**
 * Task API contracts
 *
 * @spec specs/003-todo-frontend/spec.md (FR-015, FR-016, FR-017, FR-020, FR-024)
 */

import { z } from 'zod'

// ============================================================================
// Types
// ============================================================================

export type TaskStatus = z.infer<typeof TaskStatusSchema>
export type TaskPriority = z.infer<typeof TaskPrioritySchema>
export type Task = z.infer<typeof TaskSchema>
export type TaskCreateRequest = z.infer<typeof TaskCreateRequestSchema>
export type TaskUpdateRequest = z.infer<typeof TaskUpdateRequestSchema>
export type TaskListResponse = z.infer<typeof TaskListResponseSchema>
export type TaskSingleResponse = z.infer<typeof TaskSingleResponseSchema>
export type TaskDeleteResponse = z.infer<typeof TaskDeleteResponseSchema>

// ============================================================================
// Zod Schemas
// ============================================================================

export const TaskStatusSchema = z.enum(['pending', 'in-progress', 'completed', 'deleted'])

export const TaskPrioritySchema = z.enum(['low', 'medium', 'high'])

export const TaskSchema = z.object({
  id: z.string().uuid(),
  user_id: z.string(),
  title: z.string().min(1).max(500),
  description: z.string().max(10000).nullable(),
  status: TaskStatusSchema,
  priority: TaskPrioritySchema,
  created_at: z.string().datetime(),
  updated_at: z.string().datetime(),
})

export const TaskCreateRequestSchema = z.object({
  title: z.string()
    .min(1, 'Title is required')
    .max(500, 'Title must be less than 500 characters')
    .transform(val => val.trim()),
  description: z.string()
    .max(10000, 'Description must be less than 10,000 characters')
    .transform(val => val.trim())
    .optional(),
  priority: TaskPrioritySchema.default('medium'),
})

export const TaskUpdateRequestSchema = z.object({
  title: z.string()
    .min(1, 'Title cannot be empty')
    .max(500, 'Title must be less than 500 characters')
    .transform(val => val.trim())
    .optional(),
  description: z.string()
    .max(10000, 'Description must be less than 10,000 characters')
    .transform(val => val.trim())
    .optional(),
  status: TaskStatusSchema.optional(),
  priority: TaskPrioritySchema.optional(),
})

export const TaskListResponseSchema = z.object({
  status: z.literal('success'),
  data: z.array(TaskSchema),
  meta: z.object({
    total: z.number(),
  }),
})

export const TaskSingleResponseSchema = z.object({
  status: z.literal('success'),
  data: TaskSchema,
  message: z.string().optional(),
})

export const TaskDeleteResponseSchema = z.object({
  status: z.literal('success'),
  message: z.string(),
})

// ============================================================================
// Error Response Schema
// ============================================================================

export const TaskErrorResponseSchema = z.object({
  status: z.literal('error'),
  code: z.enum([
    'VALIDATION_ERROR',
    'NOT_FOUND',
    'FORBIDDEN',
    'INVALID_OPERATION',
    'UNAUTHORIZED',
  ]),
  message: z.string(),
  details: z.object({
    errors: z.array(z.object({
      field: z.string(),
      message: z.string(),
      type: z.string(),
    })),
  }).optional(),
})

export type TaskErrorResponse = z.infer<typeof TaskErrorResponseSchema>

// ============================================================================
// Analytics Types (computed client-side)
// ============================================================================

export interface TaskAnalytics {
  statusDistribution: {
    pending: number
    'in-progress': number
    completed: number
  }
  priorityDistribution: {
    low: number
    medium: number
    high: number
  }
  completionTrend: {
    date: string  // ISO date (YYYY-MM-DD)
    completed: number
  }[]
}

/**
 * Compute analytics from task list
 */
export function computeAnalytics(tasks: Task[]): TaskAnalytics {
  const nonDeletedTasks = tasks.filter(t => t.status !== 'deleted')

  // Status distribution
  const statusDistribution = {
    pending: nonDeletedTasks.filter(t => t.status === 'pending').length,
    'in-progress': nonDeletedTasks.filter(t => t.status === 'in-progress').length,
    completed: nonDeletedTasks.filter(t => t.status === 'completed').length,
  }

  // Priority distribution
  const priorityDistribution = {
    low: nonDeletedTasks.filter(t => t.priority === 'low').length,
    medium: nonDeletedTasks.filter(t => t.priority === 'medium').length,
    high: nonDeletedTasks.filter(t => t.priority === 'high').length,
  }

  // Completion trend (last 30 days)
  const completedTasks = tasks.filter(t => t.status === 'completed')
  const last30Days = Array.from({ length: 30 }, (_, i) => {
    const date = new Date()
    date.setDate(date.getDate() - i)
    return date.toISOString().split('T')[0]
  }).reverse()

  const completionTrend = last30Days.map(date => ({
    date,
    completed: completedTasks.filter(t =>
      t.updated_at.startsWith(date)
    ).length,
  }))

  return {
    statusDistribution,
    priorityDistribution,
    completionTrend,
  }
}
